loopd/config.go

```diff
@@ -147,3 +205,81 @@ func Validate(cfg *Config) error {
 
 	return nil
 }
+
+// getTLSConfig generates a new self signed certificate or refreshes an existing
+// one if necessary, then returns the full TLS configuration for initializing
+// a secure server interface.
+func (d *Daemon) getTLSConfig() (*tls.Config, *credentials.TransportCredentials,
+	error) {
+
+	// Ensure we create TLS key and certificate if they don't exist.
+	if !lnrpc.FileExists(d.cfg.TLSCertPath) &&
+		!lnrpc.FileExists(d.cfg.TLSKeyPath) {
+
+		log.Infof("Generating TLS certificates...")
+		err := cert.GenCertPair(
+			defaultSelfSignedOrganization, d.cfg.TLSCertPath,
+			d.cfg.TLSKeyPath, d.cfg.TLSExtraIPs,
+			d.cfg.TLSExtraDomains, d.cfg.TLSDisableAutofill,
+			cert.DefaultAutogenValidity,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		log.Infof("Done generating TLS certificates")
+	}
+
+	certData, parsedCert, err := cert.LoadCert(
+		d.cfg.TLSCertPath, d.cfg.TLSKeyPath,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// If the certificate expired or it was outdated, delete it and the TLS
+	// key and generate a new pair.
+	if time.Now().After(parsedCert.NotAfter) {
+		log.Info("TLS certificate is expired or outdated, " +
+			"generating a new one")
+
+		err := os.Remove(d.cfg.TLSCertPath)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		err = os.Remove(d.cfg.TLSKeyPath)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		log.Infof("Renewing TLS certificates...")
+		err = cert.GenCertPair(
+			defaultSelfSignedOrganization, d.cfg.TLSCertPath,
+			d.cfg.TLSKeyPath, d.cfg.TLSExtraIPs,
+			d.cfg.TLSExtraDomains, d.cfg.TLSDisableAutofill,
+			cert.DefaultAutogenValidity,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		log.Infof("Done renewing TLS certificates")
+
+		// Reload the certificate data.
+		certData, _, err = cert.LoadCert(
+			d.cfg.TLSCertPath, d.cfg.TLSKeyPath,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	tlsCfg := cert.TLSConfFromCert(certData)
+	restCreds, err := credentials.NewClientTLSFromFile(
```

Comment:
These are just two different ways to hold the same certificate. Once we're using it as a server certificate (for gRPC) and once as a client connect CA certificate (for REST). Because the REST proxy connects to gRPC and is a client.

But you are right, something's missing. The REST server also needs to serve over a TLS connection!